File: bms/go/go-gt/readMatrix/readMatrix.go
package test

import (
	"bufio"
	"strconv"
)

var ()

const ()

func readMatrix(rd *bufio.Reader, n int64) *Matrix {
	M := NewMatrix(n)
	var i, j int64
	for i = 0; i < n; i++ {
		skip(rd)
		line, _ := rd.ReadString('\n')
		for j = 0; j < n; j++ {
			line = wskip(line)
			x, p := readUint(line)
			M.Set(j, i, x)
			if p == 0 {
				panic("bad int")
			}
			line = line[p:]
		}
	}
	return M
}

func NewMatrix(n int64) (m *Matrix) {
	m = new(Matrix)
	m.N = n
	m.A = make([]int64, n*n)
	return m
}

func skip(rd *bufio.Reader) {
	var b byte = ' '
	var err error
	for b == ' ' || b == '\t' || b == '\n' {
		b, err = rd.ReadByte()
		if err != nil {
			return
		}
	}
	rd.UnreadByte()
}

func wskip(s string) string {
	for i := 0; i < len(s); i++ {
		if s[i] != ' ' && s[i] != '\t' {
			return s[i:]
		}
	}
	return ""
}

func readUint(s string) (int64, int64) {
	i := end(s)
	x, _ := strconv.ParseInt(s[:i], 10, 64)
	return int64(x), i
}

func end(s string) (i int64) {
	for i = 0; i < int64(len(s)); i++ {
		if s[i] == ' ' || s[i] == '\t' || s[i] == '\n' {
			return i
		}
	}
	return 0
}

func (m Matrix) Set(i int64, j int64, v int64) {
	m.A[i*m.N+j] = v
}

type bufio_Reader struct {

	// reader provided by the client
	// buf read and write positions

	// last byte read for UnreadByte; -1 means invalid
	// size of last rune read for UnreadRune; -1 means invalid
}

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/fail/go/go-gt/readMatrix/claude3/CAPR/readMatrix.rs


use std::io::{BufRead, BufReader, Read};

fn read_matrix(rd: &mut BufReader<impl BufRead>, n: i64) -> Box<Matrix> {
    let mut m = Box::new(Matrix::new(n));
    let mut i = 0;
    while i < n {
        skip(rd);
        let mut line = String::new();
        rd.read_line(&mut line).unwrap();
        let mut j = 0;
        while j < n {
            line = wskip(line);
            let (x, p) = read_uint(&line);
            m.set(j, i, x);
            if p == 0 {
                panic!("bad int");
            }
            line = line[p..].to_string();
            j += 1;
        }
        i += 1;
    }
    m
}

fn skip(rd: &mut BufReader<impl BufRead>) {
    let mut b = [0u8; 1];
    loop {
        match rd.read(&mut b) {
            Ok(0) => return,
            Ok(_) => {
                if b[0] != b' ' && b[0] != b'\t' && b[0] != b'\n' {
                    rd.consume(1);
                    return;
                }
            }
            Err(_) => return,
        }
    }
}

fn wskip(s: String) -> String {
    let mut i = 0;
    while i < s.len() {
        if s.as_bytes()[i] != b' ' && s.as_bytes()[i] != b'\t' {
            return s[i..].to_string();
        }
        i += 1;
    }
    "".to_string()
}

fn read_uint(s: &str) -> (i64, usize) {
    let i = end(s);
    let x = s[..i].parse::<i64>().unwrap();
    (x, i)
}

fn end(s: &str) -> usize {
    let mut i = 0;
    while i < s.len() {
        if s.as_bytes()[i] == b' ' || s.as_bytes()[i] == b'\t' || s.as_bytes()[i] == b'\n' {
            return i;
        }
        i += 1;
    }
    0
}

struct Matrix {
    n: i64,
    a: Vec<i64>,
}

impl Matrix {
    fn new(n: i64) -> Matrix {
        Matrix {
            n,
            a: vec![0; (n * n) as usize],
        }
    }

    fn set(&mut self, i: i64, j: i64, v: i64) {
        self.a[(i * self.n + j) as usize] = v;
    }
}

